lib/components/metrics.js

```diff
@@ -0,0 +1,293 @@
+    var stats = _read_proc_self_stat();
+    var starttime_sec = stats[19] / TICKS_PER_SEC;
+
+    var btime_line = fs.readFileSync("/proc/stat")
+        .toString().split(/\n/).filter((l) => l.match(/^btime /))[0];
+    var btime = Number(btime_line.split(" ")[1]);
+
+    return btime + starttime_sec;
+}
+
+Metrics.prototype.refresh = function() {
+    this._collectors.forEach((f) => f());
+};
+
+/**
+ * Adds a new collector function. These collector functions are run whenever
+ * the /metrics page is about to be generated, allowing code to update values
+ * of gauges.
+ * @param {Function} func A new collector function.
+ * This function is passed no arguments and is not expected to return anything.
+ * It runs purely to have a side-effect on previously registered gauges.
+ */
+Metrics.prototype.addCollector = function(func) {
+    this._collectors.push(func);
+};
+
+/**
+ * Adds a new gauge metric.
+ * @param {Object} opts Options
+ * @param {string} opts.namespace An optional toplevel namespace name for the
+ * new metric. Default: <code>"bridge"</code>.
+ * @param {string} opts.name The variable name for the new metric.
+ * @param {string} opts.help Descriptive help text for the new metric.
+ * @param {Function} opts.refresh An optional function to invoke to generate a
+ * new value for the gauge.
+ * If a refresh function is provided, it is invoked with the gauge as its only
+ * parameter. The function should call the <code>set()</code> method on this
+ * gauge in order to provide a new value for it.
+ * @return {Gauge} A gauge metric.
+ */
+Metrics.prototype.addGauge = function(opts) {
+    var refresh = opts.refresh;
+    var gauge = this._client.newGauge({
+        namespace: opts.namespace || "bridge",
+        name: opts.name,
+        help: opts.help,
+    });
+
+    if (opts.refresh) {
+        var refresh = opts.refresh;
+        this._collectors.push(() => refresh(gauge));
+    }
+
+    return gauge;
+};
+
+/**
+ * Adds a new counter metric
+ * @param {Object} opts Options
+ * @param {string} opts.namespace An optional toplevel namespace name for the
+ * new metric. Default: <code>"bridge"</code>.
+ * @param {string} opts.name The variable name for the new metric.
+ * @param {string} opts.help Descriptive help text for the new metric.
+ * Once created, the value of this metric can be incremented with the
+ * <code>incCounter</code> method.
+ */
+Metrics.prototype.addCounter = function(opts) {
+    this._counters[opts.name] = this._client.newCounter({
+        namespace: opts.namespace || "bridge",
+        name: opts.name,
+        help: opts.help,
+    });
+};
+
+/**
+ * Increments the value of a counter metric
+ * @param{string} name The name the metric was previously registered as.
+ * @param{Object} labels Optional object containing additional label values.
+ */
+Metrics.prototype.incCounter = function(name, labels) {
+    if (!this._counters[name]) {
+        console.log("TODO: missing metric " + name);
```

Comment:
Ohyes, that was left over from gradually introducing the code into the gitter/slack bridges.

Now throws ab806bf